src/v/cloud_topics/reconciler/reconciler.cc

```diff
@@ -177,94 +169,178 @@ ss::future<> reconciler::reconciliation_loop() {
 }
 
 ss::future<> reconciler::reconcile() {
-    auto object = co_await build_object();
+    auto staging_file_fut = co_await ss::coroutine::as_future(
+      _l1_io->create_tmp_file());
+    if (staging_file_fut.failed()) {
+        vlog(
+          lg.error,
+          "Exception creating staging file: {}",
+          staging_file_fut.get_exception());
+    }
+    auto staging_file_result = staging_file_fut.get();
+    if (!staging_file_result.has_value()) {
+        vlog(
+          lg.warn,
+          "Failed to create staging file: {}",
+          static_cast<int>(staging_file_result.error()));
+        co_return;
+    }
+
+    auto staging_file = std::move(staging_file_result.value());
+    auto output_stream = co_await staging_file->output_stream();
+
+    auto builder = l1::object_builder::create(
+      std::move(output_stream), l1::object_builder::options{});
+
+    auto object_fut = co_await ss::coroutine::as_future(
+      build_object(builder.get(), staging_file.get()));
+    co_await builder->close(); // Always.
+    if (object_fut.failed()) {
+        vlog(
+          lg.error,
+          "Exception building object: {}",
+          object_fut.get_exception());
```

Comment:
ditto Noah's comment in the other PR. Let's pull this out of the macro invocation.